File: src/tape.ts
import { base4toInt, choice } from "utils";

export type Base4 = 0 | 1 | 2 | 3;

const base4Options = [0, 1, 2, 3] as const;

export const getRandomBase4 = () => choice(base4Options)

export class Tape {
  data: Uint8Array;
  pointer: number = 0;

  static random(length: number): Tape {
    return new Tape(Uint8Array.from({ length }, () => getRandomBase4()));
  }

  constructor(data: Uint8Array) {
    if (!data.length) throw new Error("Данные ленты не могут быть пустыми.");
    this.data = data;
  }

  private read(): Base4 {
    const value = this.data[this.pointer];
    this.pointer = (this.pointer + 1) % this.data.length;
    return value as Base4;
  }

  jump(n: number): void {
    this.pointer = (this.pointer + n * 3) % this.data.length;
  }

  readInt(): number {
    const a = this.read();
    const b = this.read();
    const c = this.read();
    return base4toInt(a, b, c);
  }

  readFloat(): number {
    return this.readInt() / 64;
  }
}

File: crates/hexolution-sim/src/tape.rs
use crate::color::base4_to_int;
use crate::js_rng;

#[derive(Clone)]
pub struct Tape {
    pub data: Vec<u8>,
    pub pointer: usize,
}

impl Tape {
    pub fn random(length: usize) -> Self {
        let mut data = vec![0u8; length];
        for slot in &mut data {
            *slot = js_rng::random_base4();
        }
        Self { data, pointer: 0 }
    }

    pub fn from_data(data: Vec<u8>) -> Self {
        Self { data, pointer: 0 }
    }

    fn read(&mut self) -> u8 {
        let value = self.data[self.pointer];
        self.pointer = (self.pointer + 1) % self.data.len();
        value
    }

    pub fn jump(&mut self, n: u8) {
        self.pointer = (self.pointer + (n as usize) * 3) % self.data.len();
    }

    pub fn read_int(&mut self) -> u8 {
        let a = self.read();
        let b = self.read();
        let c = self.read();
        base4_to_int(a, b, c)
    }

    pub fn read_float(&mut self) -> f64 {
        self.read_int() as f64 / 64.0
    }

    /// Matches JS `Uint8Array.join("")`.
    pub fn hash_string(&self) -> String {
        let mut s = String::with_capacity(self.data.len());
        for &b in &self.data {
            s.push(char::from(b + b'0'));
        }
        s
    }
}
